Removed ifc::File's macro's and made the caching threadsafe.

FireFlyForLife opened the pull request:
To me this code is more readable. Following the call sites is now very straightforward instead of looking at nested macro's.

It also previously initialized the partition cache in a thread unsafe manner. By using a local static the initialization is guaranteed to be threadsafe.

FireFlyForLife committed:
Removed ifc::File's macro's and made the caching threadsafe.

FireFlyForLife committed:
Removed unnecessary include.

AndreyG commented:
It cannot work in general, only if there is single instance of `ifc::File` in a program (because static variable is a singleton).

FireFlyForLife commented:
Aah right, I didn't think about multiple instances.

So small suggestion instead:
Could I get rid of the nested macro's? Just have 1 definition and 1 declaration macro.

FireFlyForLife committed:
Reverted previous implementation with flaw.

AndreyG commented:
Do you mean leave only `DECLARE_PARTITION_GETTER` but elminitate `DECLARE_[DECL/TYPE/EXPR]_PARTITION_GETTER`?
I'm not sure why?

FireFlyForLife committed:
Simplified ifc::File partition macro's.

FireFlyForLife commented:
Yes exactly, to me it's much more readable, and easier to reason about. Take a look at how it looks in my latest update.

AndreyG commented:
I'm ok with merging `Partition` and `TypedPartition` and using `constexpr if` instead of this separation (https://github.com/AndreyG/ifc-reader/pull/28/files#diff-fbc7f4f12a626cd06400223a2dbe3cc86c63f0db66a5a503db63c88bdc118f1aR50). But I like [current](https://github.com/AndreyG/ifc-reader/pull/28/commits/fad372aba9d16306145d9aadd0adc9c10324c23f#diff-499f947d07e7d8e093cb9911338ce9d0dd52412b4cde7be1cea2bbdd304d6dd2L52-L75) declaration/definition of `File` methods more then [suggested](https://github.com/AndreyG/ifc-reader/pull/28/commits/fad372aba9d16306145d9aadd0adc9c10324c23f#diff-499f947d07e7d8e093cb9911338ce9d0dd52412b4cde7be1cea2bbdd304d6dd2R46-R64) one.

FireFlyForLife commented:
I will ask for you to reconsider. The current one adds extra mental overhead for people who don't know the codebase, on top of the fact that macro usage is considered bad practice.

You will see that in terms of length it doesn't even make a difference, you either mention `_ATTR_` in the macroname or `AttrIndex` in the parameters. The latter I strongly feel is more straightforward to understand, and is more akin to what is expected in a normal function call.

AndreyG commented:
I'm not sure about the complexity, IMO nested macro calls are no more difficult than nested function calls, but it's not that important for me. Could you, please, just align [your new code](https://github.com/AndreyG/ifc-reader/pull/28/commits/fad372aba9d16306145d9aadd0adc9c10324c23f#diff-499f947d07e7d8e093cb9911338ce9d0dd52412b4cde7be1cea2bbdd304d6dd2R46-R64) by columns as was done for [the current one](https://github.com/AndreyG/ifc-reader/pull/28/commits/fad372aba9d16306145d9aadd0adc9c10324c23f#diff-499f947d07e7d8e093cb9911338ce9d0dd52412b4cde7be1cea2bbdd304d6dd2L52-L75).

FireFlyForLife committed:
Column aligned partition getters.
 - Ordered File.cpp partitions like File.h
 - Fixed some misc functions not using cached partitions.

FireFlyForLife commented:
Thanks a lot for considering the change! To me it's a lot clearer now to work with.
I've column aligned all the partition getter parameters based on which block they were in.
I've also fixed a couple functions which didn't cache their partition.
